File: test/Interop/Cxx/cxx-impl/Inputs/operators-execution.swift
import Operators

extension Vector {
  // bool Vector::operator==(const Vector &other) const;
  @unsafe @cxx(`operator==`) @implementation
  public func equals(_ other: Vector) -> Bool { return x == other.x }

  // bool Vector::operator<(const Vector &other) const;
  @unsafe @cxx(`operator<`) @implementation
  public func less(_ other: Vector) -> Bool { return x < other.x }

  // Vector Vector::operator+(const Vector &other) const;
  @unsafe @cxx(`operator+`) @implementation
  public func plus(_ other: Vector) -> Vector { return Vector(x: x + other.x) }

  // Vector Vector::operator+(int k) const;
  @cxx(`operator+`) @implementation
  public func plus(_ k: Int32) -> Vector { return Vector(x: x + k) }

  // Vector Vector::operator-() const;
  @cxx(`operator-`) @implementation
  public func negated() -> Vector { return Vector(x: -x) }

  // Vector Vector::operator-(const Vector &other) const;
  @unsafe @cxx(`operator-`) @implementation
  public func minus(_ other: Vector) -> Vector { return Vector(x: x - other.x) }

  // Vector &Vector::operator+=(const Vector &other);
  @unsafe @cxx(`operator+=`) @implementation
  public mutating func plusEquals(_ other: Vector) -> UnsafeMutablePointer<Vector> {
    x += other.x
    return withUnsafeMutablePointer(to: &self) { $0 }
  }

  // int Vector::operator[](int i) const;
  @cxx(`operator[]`) @implementation
  public func element(_ i: Int32) -> Int32 { return x + i }

  // int Vector::operator()(int i) const;
  @cxx(`operator()`) @implementation
  public func call(_ i: Int32) -> Int32 { return x * i }

  // Vector &Vector::operator++();
  @unsafe @cxx(`operator++`) @implementation
  public mutating func increment() -> UnsafeMutablePointer<Vector> {
    x += 1
    return withUnsafeMutablePointer(to: &self) { $0 }
  }

  // Vector Vector::operator++(int);
  @cxx(`operator++`) @implementation
  public mutating func postIncrement(_: Int32) -> Vector {
    let old = self
    x += 1
    return old
  }
}

// bool operator!=(const Vector &a, const Vector &b);
@unsafe @cxx @implementation
public func != (a: Vector, b: Vector) -> Bool { return a.x != b.x }

// Vector operator*(const Vector &a, int k);
@unsafe @cxx(`operator*`) @implementation
public func times(_ a: Vector, _ k: Int32) -> Vector { return Vector(x: a.x * k) }

// bool Outer::operator==(const Point &a, const Point &b);
@unsafe @cxx @implementation
public func == (a: Outer.Point, b: Outer.Point) -> Bool { return a.v == b.v }

extension Handle {
  // bool Handle::operator==(const Handle &other) const;
  @unsafe @cxx(`operator==`) @implementation
  public func equals(_ other: Handle) -> Bool { return value == other.value }
}

// bool operator<(const Handle &a, const Handle &b);
@unsafe @cxx @implementation
public func < (a: Handle, b: Handle) -> Bool { return a.value < b.value }

// int swiftCallsOperators(const Vector &a, const Vector &b);
// Swift-side uses of the imported operators reach the implementations above.
@unsafe @cxx @implementation
public func swiftCallsOperators(_ a: Vector, _ b: Vector) -> Int32 {
  var v = a
  let w = b
  var n: Int32 = 0
  if v == w { n += 1 }
  if v != w { n += 10 }
  if v < w { n += 100 }
  v += w
  n += 1000 * v.x
  n += 10000 * (v + w).x
  n += 100000 * (-v).x
  return n
}

File: test/Interop/Cxx/cxx-impl/operators-execution.cpp
// Executable end-to-end test: a C++ program uses operators declared in
// operators.h whose bodies are provided in Swift via `@cxx @implementation`.

// RUN: %empty-directory(%t)
// RUN: %target-interop-build-clangxx \
// RUN:   -c %s \
// RUN:   -I %S/Inputs \
// RUN:   -o %t/operators-execution-main.o
// RUN: %target-interop-build-swift \
// RUN:   -enable-experimental-feature CxxImplementation \
// RUN:   -Xfrontend -disable-availability-checking \
// RUN:   -module-name OperatorsExecutionMain \
// RUN:   -parse-as-library \
// RUN:   -I %S/Inputs \
// RUN:   -Xlinker %t/operators-execution-main.o \
// RUN:   %S/Inputs/operators-execution.swift \
// RUN:   -o %t/operators-execution
// RUN: %target-codesign %t/operators-execution
// RUN: %target-run %t/operators-execution | %FileCheck %s

// REQUIRES: executable_test
// REQUIRES: swift_feature_CxxImplementation

#include <stdio.h>

#include "operators.h"

// Retains minus releases.
static int liveHandles = 0;

void retainHandle(Handle *) { ++liveHandles; }
void releaseHandle(Handle *) { --liveHandles; }

int main() {
  Vector a{3}, b{4}, c{3};

  printf("equals=%d %d\n", a == b, a == c);
  // CHECK: equals=0 1

  printf("less=%d %d\n", a < b, b < a);
  // CHECK: less=1 0

  Vector plusVector = a + b;
  Vector plusInt = a + 10;
  printf("plus=%d %d\n", plusVector.x, plusInt.x);
  // CHECK: plus=7 13

  Vector negated = -a;
  Vector minus = a - b;
  printf("minus=%d %d\n", negated.x, minus.x);
  // CHECK: minus=-3 -1

  // The compound assignment returns its receiver.
  Vector &plusEqualsRes = (a += b);
  printf("plusEquals=%d same=%d\n", a.x, &plusEqualsRes == &a);
  // CHECK: plusEquals=7 same=1

  printf("element=%d call=%d\n", a[5], a(2));
  // CHECK: element=12 call=14

  Vector &incrementRes = ++a;
  printf("increment=%d same=%d\n", a.x, &incrementRes == &a);
  // CHECK: increment=8 same=1

  Vector postIncrementRes = a++;
  printf("postIncrement=%d %d\n", postIncrementRes.x, a.x);
  // CHECK: postIncrement=8 9

  printf("notEqual=%d %d\n", a != b, b != b);
  // CHECK: notEqual=1 0

  Vector times = a * 2;
  printf("times=%d\n", times.x);
  // CHECK: times=18

  Outer::Point p{1}, q{1}, r{2};
  printf("point=%d %d\n", p == q, p == r);
  // CHECK: point=1 0

  Handle h{5}, k{7};
  int handleEquals = h == k;
  int handleEqualsSelf = h == h;
  int handleLess = h < k;
  printf("handle=%d %d %d live=%d\n", handleEquals, handleEqualsSelf,
         handleLess, liveHandles);
  // CHECK: handle=0 1 1 live=0

  Vector d{3}, e{4};
  printf("swiftCallsOperators=%d\n", swiftCallsOperators(d, e));
  // CHECK: swiftCallsOperators=-582890
  return 0;
}
